src/contractor/contractor.cpp

```diff
@@ -75,6 +75,12 @@ int Contractor::Run()
     EdgeID number_of_edge_based_nodes = updater.LoadAndUpdateEdgeExpandedGraph(
         edge_based_edge_list, node_weights, connectivity_checksum);
 
+    // Convert node weights for oneway streets to INVALID_EDGE_WEIGHT
+    for (auto &weight : node_weights)
+    {
+        weight = (weight & 0x80000000) ? INVALID_EDGE_WEIGHT : weight;
```

Comment:
I don't understand this part? We change the weight to INVALID_EDGE_WEIGHT if we MSB is set, but why not save it as `INVALID_EDGE_WEIGHT` in the first place?

Reply:
In the extractor where the value is computed it is not possible to say how node weights will be used later:
* in contraction it must be INVALID_EDGE_WEIGHT to make loop edges at https://github.com/Project-OSRM/osrm-backend/blob/3ee94959eab7aacb314fc68220a67d842438bd13/src/contractor/graph_contractor.cpp#L193 If `node_weights[node]` is not `INVALID_EDGE_WEIGHT` then `path_weight` of the loop edge will be always larger than the node weight value and the condition will be always false.

* for MLD `node_weights[node]` must be a correct value  otherwise `GetNodeWeight` will return `INVALID_EDGE_WEIGHT` that will break routing

The MSB flag is used to distinguish these usage scenarios.

Reply:
Oh the problem is the reverse search of MLD?  Okay yeah that is pretty ugly, but makes sense. I feel this deserves a better comment because this is a "hack".